File: app/src/main/java/com/babestudios/companyinfouk/CompaniesHouseApplication.java
package com.babestudios.companyinfouk;

import android.app.Application;
import android.content.Context;

import com.babestudios.companyinfouk.injection.ApplicationComponent;
import com.babestudios.companyinfouk.injection.ApplicationModule;
import com.babestudios.companyinfouk.injection.DaggerApplicationComponent;
import com.google.firebase.analytics.FirebaseAnalytics;

public class CompaniesHouseApplication extends Application {

	private static CompaniesHouseApplication instance;

	private FirebaseAnalytics firebaseAnalytics;

	public CompaniesHouseApplication() {
		instance = this;
	}

	ApplicationComponent mApplicationComponent;

	@Override
	public void onCreate() {
		super.onCreate();
		firebaseAnalytics = FirebaseAnalytics.getInstance(this);
		logAppOpen();
		mApplicationComponent = DaggerApplicationComponent.builder()
				.applicationModule(new ApplicationModule(this))
				.build();
		mApplicationComponent.inject(this);
	}

	public static Context getContext() {
		return instance;
	}

	public ApplicationComponent getApplicationComponent() {
		return mApplicationComponent;
	}

	public static CompaniesHouseApplication getInstance() {
		return instance;
	}

	public FirebaseAnalytics getFirebaseAnalytics() {
		return firebaseAnalytics;
	}

	private void logAppOpen(){
		firebaseAnalytics.logEvent(FirebaseAnalytics.Event.APP_OPEN, null);

	}


}

File: app/src/main/java/com/babestudios/companyinfouk/CompaniesHouseApplication.kt
package com.babestudios.companyinfouk

import android.app.Application
import com.babestudios.companyinfouk.shared.data.androidDataModule
import com.babestudios.companyinfouk.shared.di.coroutineContextModule
import com.babestudios.companyinfouk.shared.di.dataModule
import com.babestudios.companyinfouk.shared.domain.api.CompaniesRepository
import org.koin.android.ext.koin.androidContext
import org.koin.core.component.KoinComponent
import org.koin.core.context.startKoin

open class CompaniesHouseApplication : Application(), KoinComponent {

	// @Inject
	// lateinit var companiesRepository: CompaniesRepository

	// private var currentActivity: AppCompatActivity? = null

	private lateinit var companiesRepository: CompaniesRepository

	override fun onCreate() {
		super.onCreate()
		// if (BuildConfig.DEBUG) {
		// 	plant(Timber.DebugTree())
		// }

		startKoin {
			androidContext(this@CompaniesHouseApplication)
			modules(androidDataModule, dataModule, coroutineContextModule)
		}

		companiesRepository = getKoin().get()

		logAppOpen(companiesRepository)
	}

	private fun logAppOpen(companiesRepository: CompaniesRepository) {
		companiesRepository.logAppOpen()
	}

	// override fun getCurrentActivity(): AppCompatActivity? {
	// 	return currentActivity
	// }
	//
	// override fun setCurrentActivity(mCurrentActivity: AppCompatActivity) {
	// 	this.currentActivity = mCurrentActivity
	// }
}
